### PR TITLE
Implement data with named columns using `xarray`

### LINKED ISSUE #53: Implement tensor axis labeling for DIANNA completely
A few times, and again today from Matthieu, the issue came up that getting the tensor axes in the correct order is a huge annoyance and/or confuses users. The preprocessing function provides a way to fix it, but it is still difficult to wrap one's head around.

To make not only DIANNA but really all of tensor-based deep learning more convenient, it would help if we had a tensor type with annotated axes. Something that acts like

```python
class AnnotatedAxesTensor:
    def __init__(self, data, axis_labels):
        ...

batch_size = 20
channels = 38
fidgets = 10
worbles = 204
sigwatts = 2

data = np.zeros((batch_size, channels, fidgets, worbles, sigwatts))

t = AnnotatedAxesTensor(data, ["batch", "channel", "fidget", "worble", "sigwatt"])

print(t.batch[0])
```

which would output the first batch. Models that use these tensors will be able to "transpose" / rotate axes in the underlying tensor automatically, which will make the user input side a lot easier. It can also give a warning when it does indeed need to reshape the underlying tensor, because this may impact performance, so the user may want to reshape beforehand, but now it will be easy, because the implementation warning can just say exactly how they should do the reshaping:

```python
print("please reshape for optimal performance: t.rotate_axes["batch", "sigwatt", "worble" ...]")
```

I realize that this may be out of scope for the DIANNA project currently, but maybe this is something to look into in the future, if you think it would be useful.

### PR DESCRIPTION
**Pull request by loostrum**
Added utilities to create an xarray object and move axes around.
The new feature is used in RISE.

To-Do:
* ~~implement in LIME~~

Fixes #53

### CODE REVIEW

**Comment by loostrum** [comment]
Now implemented for LIME on images as well.

Note: The lime_filename test originally used a numpy axis transpose to get the data in the right shape, but there also swapped the y and x axis of the fake data. The updated tests uses named axes and no longer swaps y and x, hence the reference output needed to be updated as well. 

Before update the reference output, I verified that the test passed when manually swapping those axes back

File: tests/test_lime.py [hunk 1]
```diff
@@ -11,30 +11,35 @@ class LimeOnImages(TestCase):
...
         assert heatmap.shape == input_data[0].shape[:2]
         assert np.allclose(heatmap, heatmap_expected, atol=.01)
 
     def test_lime_filename(self):
         np.random.seed(42)
         model_filename = 'tests/test_data/mnist_model.onnx'
-        black_and_white = generate_data(batch_size=1).transpose((0, 3, 2, 1))
-        input_data = np.zeros(list(black_and_white.shape[:-1]) + [3]) + black_and_white
+
+        # batch channel y x
```
Comment (egpbos) [inline tests/test_lime.py:27]: This comment can be left out, now that it is written in code :)

File: tests/test_rise.py [hunk 2]
```diff
@@ -15,19 +15,20 @@ class RiseOnImages(TestCase):
     def test_rise_function(self):
         # shape is batch, y, x, channel
```
Comment (egpbos) [inline tests/test_rise.py:16]: Remove comment and ideally add the y and x labels to the `axes_labels` as well. They may not actually be useful in this case (that you could note in a comment), but it's still good practice to have descriptive data, I think.

File: tests/test_rise.py [hunk 3]
```diff
@@ -15,19 +15,20 @@ class RiseOnImages(TestCase):
...
+        heatmaps = dianna.explain_image(run_model, input_data, method="RISE", axes_labels=axes_labels, n_masks=200)
 
         assert heatmaps[0].shape == input_data[0].shape[:2]
 
     def test_rise_filename(self):
         model_filename = 'tests/test_data/mnist_model.onnx'
-        # original shape is batch, channel, y, x
-        input_data = generate_data(batch_size=1).transpose((0, 2, 3, 1)).astype(np.float32)
+        input_data = generate_data(batch_size=1).astype(np.float32)
+        axes_labels = {0: 'batch', 1: 'channels'}
```
Comment (egpbos) [inline tests/test_rise.py:27]: y, x here as well then.

**Review by egpbos — CHANGES_REQUESTED** [review]
Looks great, very nice improvement! I made a few suggestions, feel free to say no if you disagree. After handling those, I think it can be merged.

File: dianna/methods/lime.py [hunk 4]
```diff
@@ -149,3 +154,48 @@ def explain_image(self,
         mask = explanation.get_image_and_mask(label, positive_only=positive_only, hide_rest=hide_rest,
                                               num_features=num_features, **get_image_and_mask_kwargs)[1]
         return mask
+
+    def _prepare_image_data(self, input_data):
+        """
+        Transforms the data to be of the shape and type LIME expects
+
+        Args:
+            input_data (np.array): Data to be explained
```
[the remark this answers is not in the document]
Comment (loostrum) [reply dianna/methods/lime.py:163]: Numpy-compatible indeed, I changed it

File: dianna/methods/lime.py [hunk 5]
```diff
@@ -56,6 +61,8 @@ def __init__(self,
                                                   )
 
         self.preprocess_function = preprocess_function
+        self.axes_labels = axes_labels if axes_labels is not None else []
+        self.required_labels = ('batch', 'channels')
```
[the remark this answers is not in the document]
Comment (loostrum) [reply dianna/methods/lime.py:65]: Good suggestion! I'll change that in both LIME and RISE

File: dianna/methods/lime.py [hunk 6]
```diff
@@ -149,3 +154,48 @@ def explain_image(self,
...
+        Returns:
+            Function that first ensures the data has the same shape and type as the input data,
+            then runs the users' preprocessing function
+        """
+        # LIME generates numpy arrays, so it is easiest here to use np.transpose to put the channels
+        # axis in the right place
+        # one is added to the channels axis index because there is an extra first axis: the batch axis
+        if self.preprocess_function is None:
+            return lambda data: np.moveaxis(data, -1, channel_axis_index+1).astype(dtype)
+        return lambda data: self.preprocess_function(np.moveaxis(data, -1, channel_axis_index+1).astype(dtype))
```
[the remark this answers is not in the document]
Comment (loostrum) [reply dianna/methods/lime.py]: 👍

File: tests/test_rise.py (hunk 2 above)
Comment (loostrum) [reply tests/test_rise.py:16]: Agreed

File: dianna/methods/lime.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (loostrum) [reply dianna/methods/lime.py]: There's this supernice "commit suggestion" button but it won't work :/ 
I'll just copy-paste it to my own machine and commit from there

[the remark this answers is not in the document]
Comment (loostrum) [reply dianna/methods/lime.py]: Note I specifically removed the `else` because the linter complains about it (redundant because of the return)

**Comment by loostrum** [comment]
Suggestions implemented, ready for another round of review

File: dianna/methods/lime.py [hunk 7]
```diff
@@ -149,3 +155,51 @@ def explain_image(self,
...
+
+        Returns:
+            Function that first ensures the data has the same shape and type as the input data,
+            then runs the users' preprocessing function
+        """
+        # LIME generates numpy arrays, so it is easiest here to use np.transpose to put the channels
+        # axis in the right place
+        # one is added to the channels axis index because there is an extra first axis: the batch axis
+        def moveaxis_function(data):
+            return np.moveaxis(data, -1, channel_axis_index+1).astype(dtype)
```
Comment (egpbos) [inline dianna/methods/lime.py:201]:
```suggestion
            return np.moveaxis(data, -1, channel_axis_index + 1).astype(dtype)
```

Again the missing whitespace.

**Review by egpbos — CHANGES_REQUESTED** [review]
Mini thing, other than that LGTM!

**Review by egpbos — APPROVED** [review]

**Pull request merged**